src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -222,14 +252,133 @@ public static String encode(String s, String enc)
     public static String encode(String s, Charset charset) {
         Objects.requireNonNull(charset, "charset");
 
+        int spaceCount = 0;
+        int needEncodingCount = 0;
+        boolean utf8 = charset == StandardCharsets.UTF_8;
+        int ut8Length = 0;
+        boolean surrogateError = false;
+        for (int i = 0, length = s.length(); i < length; ++i) {
+            char c = s.charAt(i);
+            if (c < 0x80) {
+                if (c == ' ') {
+                    spaceCount++;
+                }
+                if (!dotNeedEncoding(c)) {
+                    needEncodingCount++;
+                }
+            } else {
+                needEncodingCount++;
+                if (utf8) {
+                    if (c < 0x800) {
+                        // 2 bytes
+                        ut8Length += 3;
+                    } else if (Character.isSurrogate(c)) {
+                        // Have a surrogate pair
+                        if (Character.isHighSurrogate(c) && i + 1 < length) {
+                            char d = s.charAt(i + 1);
+                            if (Character.isLowSurrogate(d)) {
+                                ut8Length += 8;
+                                i++;
+                                continue;
+                            }
+                        }
+                        surrogateError = true;
+                        ut8Length += 3;
+                    } else {
+                        // 3 bytes
+                        ut8Length += 6;
+                    }
+                }
+            }
+        }
+
+        if (needEncodingCount == 0 && spaceCount == 0) {
+            // not need change
+            return s;
+        }
+
+        if (utf8 & !surrogateError) {
+            return encodeUTF8(s, needEncodingCount, ut8Length);
+        }
+
+        return encodeSlow(s, charset);
+    }
+
+    private static String encodeUTF8(String s, int needEncodingCount, int ut8Length) {
+        byte[] buf = new byte[s.length() + needEncodingCount * 2 + ut8Length];
+
+        int off = 0;
+        for (int i = 0, length = s.length(); i < length; ++i) {
```

Comment:
Can move the `length` definition to before `buf` assignment so we use `length` instead of `s.length()`